coordinator/src/notification_service.rs

```diff
@@ -0,0 +1,132 @@
+    ///
+    /// If on empty string is passed in the constructor, the service will not send any notification.
+    /// It will only log the notification that it would have sent.
+    pub fn new(fcm_api_key: String) -> Self {
+        if fcm_api_key.is_empty() {
+            // Lot it as error, as it production it should always be set
+            tracing::error!("FCM API key is empty. No notifications will not be sent.");
+        }
+
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        // TODO: use RAII here
+        tokio::spawn({
+            let fcm_api_key = fcm_api_key;
+            let client = fcm::Client::new();
+            async move {
+                while let Some(Notification {
+                    user_fcm_token,
+                    notification_kind,
+                }) = notification_receiver.recv().await
+                {
+                    tracing::info!("Sending notification: {notification_kind} to {user_fcm_token}");
+
+                    if !fcm_api_key.is_empty() {
+                        let notification = build_notification(notification_kind);
+                        if let Err(e) =
+                            send_notification(&client, &fcm_api_key, &user_fcm_token, notification)
+                                .await
+                        {
+                            tracing::error!("Could not send notification to FCM: {:?}", e);
+                        }
+                    }
+                }
+            }
+        });
+
+        Self {
+            notification_sender,
+        }
+    }
+
+    /// Constructs a new sender. Use a sender to send notification from any part of the system.
+    pub fn get_sender(&self) -> mpsc::Sender<Notification> {
+        self.notification_sender.clone()
+    }
+}
+
+/// Prepares the notification text
+fn build_notification<'a>(kind: NotificationKind) -> fcm::Notification<'a> {
+    let mut notification_builder = fcm::NotificationBuilder::new();
+    match kind {
+        NotificationKind::ChannelClose => {
+            notification_builder.body("Close channel request.");
+            notification_builder.title("Someone wants to close a position with you! 🌻");
+        }
+        NotificationKind::PositionSoonToExpire => {
+            notification_builder.title("Your Position is about to expire");
+            notification_builder.body("Open the app to react.");
+        }
+        NotificationKind::PositionExpire => {
+            notification_builder.title("Your position has expired");
+            notification_builder.body("Open the app to react.");
+        }
+    }
+    notification_builder.finalize()
+}
+
+async fn send_notification<'a>(
+    client: &fcm::Client,
+    api_key: &str,
+    fcm_token: &str,
+    notification: fcm::Notification<'a>,
+) -> Result<()> {
+    anyhow::ensure!(!api_key.is_empty(), "FCM API key is empty");
+
+    let mut message_builder = fcm::MessageBuilder::new(api_key, fcm_token);
```

Comment:
ah, here is `fcm` used. I would add the dependency in the same commit to make it atomic. 
For example, if we were to undo this commit only we would miss removing `fcm` from cargo.toml.